### DISCUSSION TITLE
connectField vs useField

### DISCUSSION THREAD

**Discussion by simplecommerce** [vazco/uniforms#1079] [General]
Hi I had a quick question.

I know that useField is a hook version of connectField, but in this case, would there be a reason to still use connectField vs useField for newer projects going in the future?

I tested both in my current project and they do the same, I am just trying to find out if I would have a reason to keep using connectField instead of useField from now on.

Thanks!

**Comment by radekmie** [comment]
Hi @simplecommerce! That's a very good question! (It's actually _too_ good and you shouldn't have to ask it in the first place... We'll add it to the docs later.)

From the first version of uniforms, we've supported the `connectField` higher order component. For a long time (pre-v3), it created a class component that utilized the old React Context API and provided the wrapped component with all uniforms-related props.

Starting in v3, we introduced the `useField` hook. The main difference is that it utilizes the new React Context API, which has some significant (for us) consequences. The main one is that to change the childrens' context, you have to render a `Provider` component. (You can read more about it in #405).

Starting in v3, `connectField` no longer creates a class component. It creates a function component that uses the `useHook`, and renders the `Provider` component if needed (check [the implementation](https://github.com/vazco/uniforms/blob/80b6a389a1f6271e844bf350a2b6f22592a7fcf6/packages/uniforms/src/connectField.tsx#L55-L57)). So, to sum up, the difference is that `connectField` _may_ render the `Provider` component, so the difference is relevant only to fields that may contain other fields, like `NestField` or `ListField`.

**Reply by simplecommerce** [reply]
@radekmie Thanks for explaining, so if I understand correctly, if I use the `useField` from now on, I shouldn't have any reasons to use the `connectField` anymore correct? Or are there any specific cases/scenarios where I could need to use `connectField` because `useField` wouldn't work? I can't think of any.

**Reply by radekmie** [reply]
As long as you're creating a field that has no nested fields, `useField` will be more than enough. But if you'll be creating a field with nested ones (e.g., an entire custom theme), then `connectField` it is.

**Reply by simplecommerce** [reply]
@radekmie  Thank you!
